.Net: Expose the token counter in the TextChunker

anthonypuppo opened the pull request:
### Motivation and Context

The TextChunker currently uses a rough estimate of the total number of tokens in a string (input divided by 4). This PR aims to expose the counter logic to be caller configurable, enabling accurate token counts based on which model will be consuming the chunks.

Fixes #2146 

### Description

Adds an optional token counter function parameter to the TextChunker methods. If not supplied, it falls back to the base token counter logic.

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

shawncal commented on dotnet/src/SemanticKernel/Text/TextChunker.cs:
```diff
@@ -259,10 +273,8 @@ private static (List<string>, bool) Split(ReadOnlySpan<char> input, string? inpu
     }
 
-    private static int TokenCount(int inputLength)
+    private static int FallbackTokenCounter(string input)
```
Can you please rename this to "DefaultTokenCounter"?

anthonypuppo replied:
Done! Thanks for the tip, that method name is much better than "Fallback".

shawncal commented:
Cool change. I'll take a closer look in a bit. Also requesting review from @RogerBarreto

lemillermicrosoft commented on dotnet/src/SemanticKernel/Text/TextChunker.cs:
```diff
@@ -87,14 +101,14 @@ private static List<string> InternalSplitTextParagraphs(List<string> lines, int
             var secondLastParagraph = paragraphs[paragraphs.Count - 2];
 
-            if (TokenCount(lastParagraph.Length) < adjustedMaxTokensPerParagraph / 4)
+            if (tokenCounter(lastParagraph) < adjustedMaxTokensPerParagraph / 4)
```
I'm curious now if we were ever supposed to be doing the division by 4 here -- I can't recall the reasoning....

anthonypuppo replied:
It's a rough check to see if the last paragraph is too short (less than 1/4 of the maximum tokens per paragraph) and if so, validates and merges it with the second to last paragraph. Best to leave as is and potentially refactor once the TextChunker is enhanced with some more configuration options?

lemillermicrosoft commented on dotnet/src/SemanticKernel/Text/TextChunker.cs:
```diff
@@ -59,13 +70,16 @@ public static List<string> SplitPlainTextParagraphs(List<string> lines, int maxT
     /// <param name="lines">Lines of text.</param>
     /// <param name="maxTokensPerParagraph">Maximum number of tokens per paragraph.</param>
     /// <param name="overlapTokens">Number of tokens to overlap between paragraphs.</param>
+    /// <param name="tokenCounter">Function to count tokens in a string. If not supplied, the default counter (input length divided by 4) will be used.</param>
```
nit: I don't love leaving the implmentation details in the comments as it could change and become out of sync.

anthonypuppo replied:
Good point, removed.

lemillermicrosoft commented on dotnet/src/SemanticKernel/Text/TextChunker.cs:
```diff
@@ -129,14 +143,14 @@ private static List<string> InternalSplitTextParagraphs(List<string> lines, int
         foreach (string line in truncatedLines)
         {
-            if (paragraphBuilder.Length > 0 && TokenCount(paragraphBuilder.Length) + TokenCount(line.Length) + 1 >= maxTokensPerParagraph)
+            if (paragraphBuilder.Length > 0 && tokenCounter(paragraphBuilder.ToString()) + tokenCounter(line) + 1 >= maxTokensPerParagraph)
```
nit: You could wrap the delegate with a couple internal methods (see the Encode method support in GPT3Tokenizer).

anthonypuppo replied:
I think the GPT3Tokenizer Encode methods differ slightly in that they're publicly exposed and all call the same base method (not utilized again internally). These internal methods would need to be called quite a bit which would make even basic calls decently verbose?

Something like this...
```
private static int CountTokens(string input, TokenCounter tokenCounter) => tokenCounter(input);
private static int CountTokens(StringBuilder stringBuilder, TokenCounter tokenCounter) => tokenCounter(stringBuilder.ToString());
```

Seems overly complex for the gain.

lemillermicrosoft reviewed (commented):
A few comments. Also, it'd be nice to add or update the KernelSyntaxExamples to show how this could be used with the GPT3Tokenizer as an example. 

I'd love to add this to ConversationSummarySkill which uses the TextChunker-- but I don't yet have a great answer on how we can get the model being used from the SKContext to know which tokenizer to use. cc @shawncal

lemillermicrosoft commented:
One more thought: Please add some UTs to dotnet\src\SemanticKernel.UnitTests\Text\TextChunkerTests.cs

anthonypuppo commented:
> Also, it'd be nice to add or update the KernelSyntaxExamples to show how this could be used with the GPT3Tokenizer as an example.

Added as example 55.

> One more thought: Please add some UTs to dotnet\src\SemanticKernel.UnitTests\Text\TextChunkerTests.cs

Added a test for each method overload. Assume the extras are redundant, but I can add if you think it's necessary.

lemillermicrosoft commented on dotnet/samples/KernelSyntaxExamples/Example55_TextChunker.cs:
```diff
@@ -0,0 +1,77 @@
+
+    private static int TokenCounter(string input)
+    {
+        var tokens = GPT3Tokenizer.Encode(input);
```
<3

lemillermicrosoft reviewed (approved):
Awesome, thanks for the contribution!
